### PR TITLE
128 extend controller to handle priorities

### PR DESCRIPTION
**Pull request by samvanderzwan**
The controller can now set the sources based on their priority. However the priority cannot yet be loaded form esdl since it does not know an attribute priority for producers. We will need to discuss with Edwin and Ewoud if this can be added.
Reviewing and merging of this branch is on hold until other PRs are handled. Then this branch will need some adjustments to be able to merge with master.

### CODE REVIEW

**Comment by samvanderzwan** [comment]
This PR can also be reviewed, I also rebased it on current master branch

File: ci/win32/profiling.cmd
Comment (MichielTukker) [inline ci/win32/profiling.cmd:1]: OK  to delete,  but what does this have to do with "Extend controller to handle priorities?"

Also,  it seems like a useful tool?

File: doc/controller/controller.rst [hunk 1]
```diff
@@ -1,3 +1,31 @@
-Controller class
+Controller
 =====================================
-To be filled.
+The controller is responsible for providing set points for the assets in the simulator.
+The controller parses the ESDL and creates objects of the assets it is providing set points for.
```
Comment (MichielTukker) [inline doc/controller/controller.rst:4]: The controller parses ESDL?    The ESDL is parsed by ```simulator-core```  to create an object model of the ESDL which is used by our solver.  This parser now also creates the controller objects

File: doc/controller/controller.rst [hunk 2]
```diff
@@ -1,3 +1,31 @@
...
+When a time step is calculated the controller passes back a dict of the set points
+for the controllable assets in the network. The key of this dict is the id of the asset.
+The value is another dict. The key of this dict is the property which needs to be set
+(e.g. supply temperature, heat demand). The value is the set point for this property.
+The controller works based on the priority of the source. The controller will first allocate
+The controller will also  if the sources with priority 1 have enough capacity to provide the demand.
```
Comment (MichielTukker) [inline doc/controller/controller.rst:10]: Sentences are incomplete

File: doc/controller/controller.rst [hunk 3]
```diff
@@ -1,3 +1,31 @@
...
+if more capacity is required, the sources with priority 2 will be used etc. If the demand is lower
+then the available capacity of the observed sources, the remaining demand will be equally distributed
+over the source at the observed priority.
```
Comment (MichielTukker) [inline doc/controller/controller.rst:13]: So now we're implementing a very simple algorithm for the controller.   Does the current architecture allow for alternative control strategies?

File: src/simulator_core/entities/assets/controller/controller_classes.py
Comment (MichielTukker) [inline src/simulator_core/entities/assets/controller/controller_classes.py:1]: File should be renamed to controller_abstract right?

File: src/simulator_core/entities/assets/controller/controller_consumer.py [hunk 4]
```diff
@@ -58,24 +49,24 @@ def get_heat_demand(self, time: datetime.datetime) -> float:
         for index in range(self.start_index, len(self.profile)):
             if abs((self.profile["date"][index].to_pydatetime() - time).total_seconds()) < 3600:
                 self.start_index = index
-                return float(self.profile["values"][index])
+                if self.profile["values"][index] > self.max_power:
+                    # TODO need to pass a message that power is insufficient and is
+                    #  capped to the max power available
```
Comment (MichielTukker) [inline src/simulator_core/entities/assets/controller/controller_consumer.py:54]: Write the message to logging?

File: src/simulator_core/entities/network_controller.py [hunk 5]
```diff
@@ -40,28 +41,84 @@ def run_time_step(self, time: datetime.datetime) -> dict:
...
+            producers[source.id] = {PROPERTY_HEAT_DEMAND: source.power,
+                                    PROPERTY_TEMPERATURE_RETURN: source.temperature_return,
+                                    PROPERTY_TEMPERATURE_SUPPLY: source.temperature_supply,
+                                    PROPERTY_SET_PRESSURE: False}
+        # setting the first producer to set the pressure.
+        producers[self.producers[0].id][PROPERTY_SET_PRESSURE] = True
+        return producers
+
+    def _set_consumer_capped(self, time: datetime.datetime) -> dict:
+        """Method to set the consumer to the capped to the max avaialbl power of the producers."""
```
Comment (MichielTukker) [inline src/simulator_core/entities/network_controller.py:85]: Spelling errors  +  undocumented function param

File: src/simulator_core/entities/network_controller.py [hunk 6]
```diff
@@ -40,28 +41,84 @@ def run_time_step(self, time: datetime.datetime) -> dict:
...
+        factor = self.get_total_supply() / self.get_total_demand(time)
+        return self._set_consumer_to_demand(time=time, factor=factor)
+
+    def _set_consumer_to_demand(self, time: datetime.datetime, factor: float = 1.0) -> dict:
+        """Method to set the consumer to the demand."""
```
Comment (MichielTukker) [inline src/simulator_core/entities/network_controller.py:90]: undocumented function parameters

File: src/simulator_core/entities/network_controller.py [hunk 7]
```diff
@@ -40,28 +41,84 @@ def run_time_step(self, time: datetime.datetime) -> dict:
...
+        consumers = {}
+        for consumer in self.consumers:
+            consumers[consumer.id] = {PROPERTY_HEAT_DEMAND: consumer.get_heat_demand(time) * factor,
+                                      PROPERTY_TEMPERATURE_RETURN: consumer.temperature_return,
+                                      PROPERTY_TEMPERATURE_SUPPLY: consumer.temperature_supply}
+        return consumers
+
+    def _set_producers_based_on_priority(self, time: datetime.datetime) -> dict:
+        """Method to set the producers based on priority."""
```
Comment (MichielTukker) [inline src/simulator_core/entities/network_controller.py:99]: Undocumented function param

File: unit_test/entities/controller/test_controller_class.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    def test_run_time_step(self):
+        """Test to run the time step."""
+        # Act
+        result = self.controller.run_time_step(datetime.now())
+
+        # Assert
+        self.assertIn(self.producer1.id, result)
+        self.assertIn(self.producer2.id, result)
+        self.assertIn(self.consumer1.id, result)
+        self.assertIn(self.consumer2.id, result)
```
Comment (MichielTukker) [inline unit_test/entities/controller/test_controller_class.py:144]: Why not check the results?   this will be very valuable when the controller is improved further in the future  or the behavior/priorities are set in ESDL  (the results should not change)

**Review by MichielTukker — CHANGES_REQUESTED** [review]

File: ci/win32/profiling.cmd
Comment (samvanderzwan) [reply ci/win32/profiling.cmd:1]: nothing therefore I removed it after I added it by accident. And is certainly useful see #68

File: doc/controller/controller.rst (hunk 1 above)
Comment (samvanderzwan) [reply doc/controller/controller.rst:4]: indeed, since the controller also needs information from the esdl

File: doc/controller/controller.rst (hunk 2 above)
Comment (samvanderzwan) [reply doc/controller/controller.rst:10]: adjusted

File: src/simulator_core/entities/assets/controller/controller_classes.py
Comment (samvanderzwan) [reply src/simulator_core/entities/assets/controller/controller_classes.py:1]: Yes, done

File: src/simulator_core/entities/network_controller.py (hunk 5 above)
Comment (samvanderzwan) [reply src/simulator_core/entities/network_controller.py:85]: fixed

File: src/simulator_core/entities/network_controller.py (hunk 6 above)
Comment (samvanderzwan) [reply src/simulator_core/entities/network_controller.py:90]: fixed

File: src/simulator_core/entities/network_controller.py (hunk 7 above)
Comment (samvanderzwan) [reply src/simulator_core/entities/network_controller.py:99]: fixed

File: unit_test/entities/controller/test_controller_class.py (hunk 8 above)
Comment (samvanderzwan) [reply unit_test/entities/controller/test_controller_class.py:144]: These are also tested in the other tests. It is usefull to add a test with actual parsing esdl and see what comes back

Comment (samvanderzwan) [reply unit_test/entities/controller/test_controller_class.py:150]: added test with laoding of esdl and then creating controller and checking result for time step

File: src/simulator_core/entities/assets/controller/controller_consumer.py (hunk 4 above)
Comment (samvanderzwan) [reply src/simulator_core/entities/assets/controller/controller_consumer.py]: doen

File: doc/controller/controller.rst (hunk 3 above)
Comment (samvanderzwan) [reply doc/controller/controller.rst]: I added an abstract class which can serve as basis for other controllers

**Comment by samvanderzwan** [comment]
Made the adjustments, but also refactored the following;
1. Added abstract controller class, which in future can be used for other controllers
2. Changed the constructor of the controller class to be called with a factory method, which returns the producers and consumers as list. This also allows for easy extension of this class.

**Review by MichielTukker — CHANGES_REQUESTED** [review]
tests don't run and still see linting errors

**Review by MichielTukker — CHANGES_REQUESTED** [review]
tests don't run and still see linting errors

File: src/simulator_core/adapter/transforms/mappers.py [hunk 9]
```diff
@@ -204,23 +210,26 @@ def to_esdl(self, entity: NetworkController) -> EsdlObject:
         """
         raise NotImplementedError("EsdlControllerMapper.to_esdl()")
 
-    def to_entity(self, model: EsdlObject) -> NetworkController:
+    def to_entity(self, model: int = 1) -> tuple[list[ControllerProducer], list[ControllerConsumer]]:
```
[the remark this answers is not in the document]
Comment (samvanderzwan) [reply src/simulator_core/adapter/transforms/mappers.py]: tests have been fixed, still need to discuss how to handle this best

File: src/simulator_core/entities/assets/controller/controller_consumer.py [hunk 10]
```diff
@@ -16,22 +16,16 @@
...
 import datetime
 
+import numpy as np
 import pandas as pd
+import logging
 from simulator_core.entities.assets.asset_defaults import (DEFAULT_TEMPERATURE,
                                                            DEFAULT_TEMPERATURE_DIFFERENCE)
+from simulator_core.entities.assets.controller.controller_abstract_clas import (
+    AssetControllerAbstract)
+from simulator_core.entities.assets.esdl_asset_object import EsdlAssetObject
```
[the remark this answers is not in the document]
Comment (samvanderzwan) [reply src/simulator_core/entities/assets/controller/controller_consumer.py:26]: is similar as we do in heatnetwork pipe asset

File: src/simulator_core/entities/assets/esdl_asset_object.py [hunk 11]
```diff
@@ -59,6 +59,41 @@ def get_profile(self) -> pd.DataFrame:
                 return get_data_from_profile(esdl_port.profile[0])
         raise ValueError("No profile found for asset: " + self.esdl_asset.name)
 
+    def get_supply_temperature(self, port_type: str) -> float:
```
[the remark this answers is not in the document]
Comment (samvanderzwan) [reply src/simulator_core/entities/assets/esdl_asset_object.py:62]: not entirely then you need to add other input argument to check on, so I decided to have two separate methods knowing we have duplicate code.

**Review by MichielTukker — APPROVED** [review]

**Review by MichielTukker — APPROVED** [review]
